### ISSUE TITLE
Support NoVNC clipboard copy paste feature.

### ISSUE THREAD

**Issue by a180285** [kubevirt/kubevirt#10306]
I use noVNC html version to connect to kubevirt VM.

And when I connect to Windows VM. Is there any way to support noVNC copy paste feature?

**Comment by victortoso** [comment]
QEMU VNC's server supports clipboard since [v6.1.0](https://gitlab.com/qemu-project/qemu/-/commit/0bf41cab93e5c72dcda717abd625698b59d9ba3e) and with libvirt that should be possible since [v8.4.0](https://listman.redhat.com/archives/libvir-list/2022-March/229493.html).

As far as I understand, based on a quick read of Gerd's [qemu-cut-paste](https://www.kraxel.org/blog/2021/05/qemu-cut-paste/), QEMU will use SPICE's guest-agent (spice-vdagent) for the actual clipboard delivery to the Guest. This should work fine with a Windows VM, but user's will need to install it.

So, to sum up:
1. Add a libvirt change to enable QEMU passing clipboard content from VNC server to the Guest
2. Installing SPICE guest agent in the Guest
3. Using a VNC client that supports copy paste, I'm not at all familiar with noVNC but I believe you that it has this feature.

I can make the patches and test this when time allows it.

**Comment by oreillymj** [comment]
Hi @victortoso  ,

I'm wondering was there ever any progress made on this issue.
I'm specifically interested in 
Add a libvirt change to enable QEMU passing clipboard content from VNC server to the Guest

Is that a code change or libvirt config?

**Comment by a180285** [comment]
BTW, linux guest copy paste is also welcome. We just use send key as a workaround.

If there is any native support. It's much better.

**Comment by victortoso** [comment]
> I'm wondering was there ever any progress made on this issue.

@oreillymj, No, but thanks for the ping. I'll give it a try this week.

> Is that a code change or libvirt config?

Libvirt needs extra configuration, yes. Should be something like

```xml
  <channel type='qemu-vdagent'>
    <target type='virtio' name='com.redhat.spice.0'/>
    <source>
      <clipboard copypaste='yes'/>
      <mouse mode='client'/>
    </source>
  </channel>
```

Some more info in the commit [libvirt@05b09f03](https://gitlab.com/libvirt/libvirt/-/commit/05b09f039ea6547efc39c2c483dbb68ad5265f13)

**Comment by victortoso** [comment]
> If there is any native support. It's much better.

@a180285 , what do you define as native support?

**Comment by victortoso** [comment]
So, https://github.com/kubevirt/kubevirt/pull/10971 implements the necessary bits in KubeVirt to allow for vnc clipboard to work. The problem is that QEMU does not support live migration of the qemu-vdagent channel, see [qemu!1380](https://gitlab.com/qemu-project/qemu/-/issues/1380).

For the moment, I'm marking the PR as Draft and we should wait for QEMU developers to add support for migration.

**Comment by victortoso** [comment]
Hi, I plan to iterate over #10971 again. There are a couple of features in KubeVirt that causes the VM to be not migratable, so there is room for this feature I think. It'll need to be disabled by default.

In the meanwhile, I tested this with sidecars and it works as well. If you are trying sidecars for the first time, please give it a look to this https://github.com/kubevirt/user-guide/pull/751 where @dharmit is including related documentation to the user-guide. You can also give suggestions/feedback :)

1. Enable Sidecar [featureGates](https://kubevirt.io/user-guide/operations/activating_feature_gates/)
2. Add this configmap: [clipboard-configmap.yam](https://gist.github.com/victortoso/a8da5e2ac967b698b85b3e8a0d33ab89#file-clipboard-configmap-yaml), it adds the necessary changes to libvirt
3. Add the following annotation to your VMI

```yaml
metadata:
  annotations:
    hooks.kubevirt.io/hookSidecars: '[{"args": ["--version", "v1alpha3"], "image":"kubevirt/sidecar-shim:20240108_99b6c4bdb",
          "configMap": {"name": "clipboard-configmap","key": "my_script.sh", "hookPath": "/usr/bin/onDefineDomain"}}]'
```

**Comment by victortoso** [comment]
KubeVirt's VNC talks with QEMU VNC server and the support for clipboard there relies on SPICE guest agent to deliver the clipboard data to the Guest. Relying on SPICE is not something we would like to encourage our users to do.

I'd suggest closing the issue as there is nothing we can do in KubeVirt to solve it.
Users can, from KubeVirt v1.2.0 and forward, use the sidecar suggestion in my previous comment if they want.

**Comment by fabiand** [comment]
Something to add is: Or teach qemu-ga to support this clippboard protocol as well.

**Comment by jjkobe** [comment]
@victortoso Hi, thank you for your effort enabling copy-paste in kubevirt via VNC. I follow your guide using Sidecar to enable copy-paste, but unfortunately it does not work. I have two questions about your test.
1. Did you build qemu with `--enable-spice`？I find the default `qemu-kvm` in launcher pod does not support `-spice`

![Image](https://github.com/user-attachments/assets/1479f5da-8277-4191-8eff-8901b0017960)

If I `virsh edit` my vm, and add <graphics type='spice'> to configuration xml, I get this error:

![Image](https://github.com/user-attachments/assets/0f0c218f-5700-4917-b034-c8d8459d8480)

2. I find that your [sidecar](https://gist.github.com/victortoso/a8da5e2ac967b698b85b3e8a0d33ab89#file-clipboard-configmap-yaml) only add a <channel type='qemu-vdagent'>.  Did you add another `<graphics type='spice'>` to libvirt configuration xml or change the `<graphics type='vnc'>` to `<graphics type='spice'>`?

[Spice manual](https://www.spice-space.org/spice-user-manual.html)  lets us to add <graphics type='spice'>

**Comment by victortoso** [comment]
Hi @jjkobe, good set of questions.

> Did you build qemu with --enable-spice？I find the default qemu-kvm in launcher pod does not support -spice

No. When I tested the sidecar that enables vnc clipboard, I did it without any changes to virt-launcher image. In centos9 we don't have `--enable-spice` set since the beginning so I'm sure it isn't needed.

> If I virsh edit my vm, and add <graphics type='spice'> to configuration xml, I get this error:

QEMU in centos9 and virt-launcher image does not have SPICE enable, that's why you see the error.

> Did you add another <graphics type='spice'> to libvirt configuration xml or change the <graphics type='vnc'> to <graphics type='spice'>?

No, because we are not using SPICE for VNC clipboard copy-paste, we are using only VNC.

---

Let me try provide a way VNC clipboard works with QEMU.

1. The delivery mechanism for copy-paste is a Guest agent, which is spice-vdagent. That is a requirement to be installed in the guest.
2. QEMU configuration uses a new virtio-port called `qemu-vdagent`. 
3. You need to use a VNC client that is capable of copy-paste. On my testing, I used tigervnc as my beloved remote-viewer uses gtk-vnc and gtk-vnc is not capable of clipboard features as of Today.

The sidecar does implement the required changes in libvirt, so all you need is a clipboard capable client (3) and spice-vdagent installed in the guest (1)

If you want more context and information, the PR https://github.com/kubevirt/kubevirt/pull/10971 has links to QEMU/libvirt commits with are usually very rich in content.

**Comment by jjkobe** [comment]
@victortoso Thank you so much! 

> The delivery mechanism for copy-paste is a Guest agent, which is spice-vdagent. That is a requirement to be installed in the guest.

I have already install spice-vdagent in my vm.

```
yum install -y qemu-guest-agent spice-vdagent
systemctl start spice-vdagent
```

> QEMU configuration uses a new virtio-port called qemu-vdagent.

I follow your guide and add the sidecar

> You need to use a VNC client that is capable of copy-paste. On my testing, I used tigervnc as my beloved remote-viewer uses gtk-vnc and gtk-vnc is not capable of clipboard features as of Today.

I think this is the root cause, and I will try some other VNC client. Of course, I will try TigerVNC first!

**Comment by hw-claudio** [comment]
as an update gtk-vnc should now have the clipboard features, with the series ending in commit:

https://gitlab.gnome.org/GNOME/gtk-vnc/-/commit/6e6d5df86cb5b6fe032573e9fd4defe88f04a733

so it should be possible to use remote-viewer as well (haven't tested it, I usually use tigervnc)
